Apply pluggable security (TLS, SST) for federated execution

Jakio815 opened the pull request:
Please check descriptions in lf-lang/reactor-c#515

Jakio815 committed:
Change head to main

Jakio815 committed:
Update runtime-version

Jakio815 committed:
Make the arduino build to skip network files copied.

Jakio815 committed:
Fix reactor-c

Jakio815 committed:
Fix file names.

Jakio815 committed:
Fix names

Jakio815 committed:
Fix communication mode enum.

Jakio815 committed:
Formatting.

Jakio815 committed:
Add SST property.

Jakio815 committed:
Add to find sst library, when using communication mode sst.

Jakio815 committed:
Add SST Path property, with get functions.

Jakio815 committed:
Add SSTGenerator.

Jakio815 committed:
Add sst generator to start generate.

Jakio815 committed:
Add auth execution script and rti execution script with -sst

Jakio815 committed:
Add federate generate script to execute with -sst option, getting the config.

Jakio815 committed:
Add some error checking thread.

Jakio815 committed:
Fix SST generator.

Jakio815 committed:
Add to get RTI's port and host name.

Jakio815 committed:
Add to get RTI config for sstGenerator.

Jakio815 committed:
Add generated script to run Auth, and generate Auth with password as main reactor name.

Jakio815 committed:
Fix auth launcher with logging.

Jakio815 committed:
Fix to not make empty line.

Jakio815 committed:
Merge branch 'networkdriver' of github.com:lf-lang/lingua-franca into sst

Jakio815 committed:
Merge branch 'networkdriver' of github.com:lf-lang/lingua-franca into sst

Jakio815 committed:
Add to find cmake when sst comm-type.

Jakio815 committed:
Add mvn build command in the lfc script

Jakio815 committed:
Add the comm type=sst build in the rti local build.

Jakio815 committed:
Add authID, encryption mode, and hmacmode

Jakio815 committed:
Fix location of auth kill command.

Jakio815 committed:
Fix launch script.
1. Change RTI job number depending if it is sst or not.
2. Pass null input to auth, or else it will not work.

Jakio815 committed:
Copy compiled jar to the lf repo.

Jakio815 committed:
Add Auth kill even when RTI successfull exits.

Jakio815 committed:
Fix properties.

Jakio815 committed:
Fix target group name to RTI

Jakio815 committed:
Move sst setup to after compilation of all federates and rti.

Jakio815 committed:
Copy sst configs, keys, and auth to the src-gen to enable deployment in a single tar ball.

Jakio815 committed:
Add -sst option to work when remote deployment.

Jakio815 committed:
Update the copied properties File usable from the remote base.

Jakio815 committed:
Minor fix on comments.

Jakio815 committed:
Update copied config files to work with remote base.

Jakio815 committed:
Add auth launch script generation when deploying on remote.

Jakio815 committed:
Add sst option in rti compile command line when remote deployment.

Jakio815 committed:
Minor fixes on Auth launch code on remote side.

Jakio815 committed:
Major fix on paths for remote deployment.

Jakio815 committed:
Add sst config generation to take the remote address of the RTI and Auth.

Jakio815 committed:
Add TLS Comm type.

Jakio815 committed:
Add cmake generator to link openssl for TLS comm type.

Jakio815 committed:
Add TLS generator, generating certificates ank keys for tls, and add launch commands.

Jakio815 committed:
Fix on path.

Jakio815 committed:
refactor: update SST generator configuration and adjust federated test suite communication settings

Jakio815 committed:
SST property type to dictionary for options.

Jakio815 committed:
Copy permanent distribution keys when using option.

Jakio815 committed:
Merge branch 'master' of github.com:lf-lang/lingua-franca into sst

Jakio815 committed:
Formatting.

Jakio815 committed:
Merge branch 'master' of github.com:lf-lang/lingua-franca into sst

Jakio815 committed:
Fix broken script

Jakio815 committed:
Fix sst config generator to use permanent distribution keys when option is on.

Jakio815 committed:
Remove openssl's placeholders && add generating TLS credentials messages.

Jakio815 committed:
Don't copy auth_certs when using permanent distribution keys.

Jakio815 committed:
Add SimpleFederatedSST tests.

Jakio815 committed:
Add sst-root-path OR SST_ROOT to be mandatory when using sst.

Jakio815 committed:
Fix CI test to pull iotauth.

Jakio815 committed:
Fix to not find sst package now. Already linked via reactor-c

Jakio815 committed:
Add reactor-c

Jakio815 committed:
Fix tags for Doxygen.

Jakio815 commented on core/src/main/java/org/lflang/federated/launcher/FedLauncherGenerator.java:
```diff
@@ -456,17 +558,6 @@ private String getRemoteLaunchCode(
             + "\" 2>&1 | tee -a "
             + logFileName
             + "; \\",
-        "    # First, check if the RTI is on the PATH",
-        "    if ! command -v RTI &> /dev/null",
-        "    then",
-        "        echo \"" + ANSI_ERROR + "RTI could not be found." + ANSI_RESET + "\"",
-        "        echo \""
-            + ANSI_ERROR
-            + "The source code can be found in org.lflang/src/lib/core/federated/RTI"
-            + ANSI_RESET
-            + "\"",
-        "        exit 1",
-        "    fi",
```
Removed this part, because this assumes RTI binary is installed on the PATH of the system. The RTI is newly compiled everytime, so I deleted this part.

Jakio815 committed:
Add DistributedCount TLS and SST options with Decentralized options to test sending actual data.

Jakio815 committed:
Update reactor-c

Jakio815 committed:
Formatting.

Jakio815 committed:
Make Auth a separate pane for the tmux option.

Jakio815 committed:
Formatting.

Jakio815 committed:
Minor fix...

Jakio815 committed:
Check if CommunicationmodeProperty is supported.

Jakio815 committed:
Formatting..

Jakio815 committed:
Minor fix and reverts.

Jakio815 committed:
Formatting

Jakio815 committed:
Update reactor-c

Jakio815 committed:
Add python target support.

Jakio815 committed:
Directly install library on CI.

Jakio815 committed:
Minor fix.

Jakio815 committed:
Fix to get or default on comm type.

Jakio815 committed:
Formatting.

Jakio815 commented on core/src/main/java/org/lflang/federated/launcher/FedLauncherGenerator.java:
```diff
@@ -441,9 +544,9 @@ private String getRemoteLaunchCode(
+            + "-------------- Federation ID: "
             + ANSI_RESET
-            + "\" >> "
+            + "\"'$FEDERATION_ID' >> "
```
Wasn't working before, just a minor bug fix.

Jakio815 committed:
ㅖropagate RTI username to rtiConfig

Jakio815 committed:
Fix path to uppercase.

Jakio815 committed:
Update reactor-c

Jakio815 committed:
Update reactor-c

Jakio815 committed:
Fix to properly kill Auth when remote launch && Fix printed comments that said RTI is exited, when it is still running.

Jakio815 commented on core/src/main/java/org/lflang/federated/launcher/FedLauncherGenerator.java:
```diff
@@ -584,13 +710,7 @@ private String getDistCode(Path remoteBase, String name, String user, String hos
             + "\" >> "
             + buildShellFileName
             + "; "
-            // The >> syntax appends stdout to a file. The 2>&1 appends stderr to the same file. tee
-            // sens to stdout and file.
-            + "echo \"rm -rf build && mkdir -p build && cd build && cmake .. && make 2>&1 | tee -a "
-            + remoteBuildLogFileName
-            + "\" >> "
-            + buildShellFileName
-            + "; "
```
Moved to block above.

Jakio815 commented on core/src/main/java/org/lflang/federated/launcher/FedLauncherGenerator.java:
```diff
@@ -191,7 +206,7 @@ public void doGenerate(List<FederateInstance> federates, RtiConfig rtiConfig) {
                 "\n",
                 "echo \""
                     + ANSI_INFO
-                    + "RTI has exited. Wait for federates to exit."
```
At this point, the RTI has not exited yet. Fixed the printed statements.

Jakio815 committed:
Formatting.

hokeun commented on .github/workflows/c-tests.yml:
```diff
@@ -54,6 +54,16 @@ jobs:
+          git clone https://github.com/iotauth/iotauth.git iotauth
+          echo "SST_ROOT=$GITHUB_WORKSPACE/iotauth" >> $GITHUB_ENV
+          git clone https://github.com/iotauth/sst-c-api.git sst-c-api
+          cd sst-c-api
```
This seems problematic because it can lead to inconsistent versions between the Auth and the C API.
This should be 
```
git submodule update --init --recursive
cd entity/c
```
to be consistent with the version of iotauth, not depending on the submodule version individually.

Jakio815 replied:
Fixed in b8efb1888fcd4e53b2ca3a89b24eff37f325ff8a

hokeun commented on core/src/main/java/org/lflang/federated/generator/FedGenerator.java:
```diff
@@ -219,6 +221,15 @@ public boolean doGenerate(Resource resource, LFGeneratorContext context) throws
     // Compile an RTI for this federation.
     buildRtiLocally(context);
 
+    // If communication mode is SST, generate configurations for SST.
```
This comment should mention both SST and TLS.

Jakio815 replied:
Fixed on 0e0719c9b3d6ab7751c54ebf97170dc1ac3d6142.

hokeun commented on core/src/main/java/org/lflang/federated/generator/FedGenerator.java:
```diff
@@ -249,12 +260,24 @@ private void buildRtiLocally(LFGeneratorContext context) {
+    configureArgs.add("-Bbuild");
+    configureArgs.add("-DCMAKE_INSTALL_PREFIX=" + fileConfig.getGenPath());
+
+    // If communication mode is SST, the RTI must be built with -DCOMM_TYPE=SST.
```
Ditto, both SST and TLS should be mentioned (see `else if`).

Jakio815 replied:
Fixed on ccc833c96d790fb0b614097a6e6a0d49035a35d0.

hokeun commented on core/src/main/java/org/lflang/federated/generator/SSTGenerator.java:
```diff
@@ -0,0 +1,832 @@
+ *
+ * @author Dongha Kim
+ */
+public class SSTGenerator {
```
This class seems to only include static methods and is not instantiated with `new`. However, looking at the usage of this class, it will make more sense to make this instantiable with configurations stored as member variables (e.g., `FederationFileConfig fileConfig`), rather than giving them every time you call each method. Please refactor both SSTGenerator and TLSGenerator classes so that they can be instantiated

Jakio815 replied:
Fixed in 6e9c4808511e1c7195fad2bd7d775a1c5dbc47ba. This was a good suggestion!

hokeun commented on core/src/main/java/org/lflang/federated/generator/TLSGenerator.java:
```diff
@@ -0,0 +1,181 @@
+ * - Generate cert+key for RTI and each federate into: fed-gen/[program]/credentials/[name]/
+ * - Copy only the needed pair into: src-gen/[entity]/credentials/
+ */
+public class TLSGenerator {
```
Same as the SSTGenerator class, please make this instantiable and change methods to non-static.

Jakio815 replied:
Fixed in b5bbed7d4aa878e65cd5c1959911c9ca5627399b

I removed all unnecessary static functions, however still kept path-utility methods (e.g., `getLocalCertPath`) as `static` because they are referenced statically by external builder/launcher classes (such as `CBuildConfig` and `FedLauncherGenerator`) that do not possess a generator instance.

hokeun commented on core/src/main/java/org/lflang/federated/launcher/FedLauncherGenerator.java:
```diff
@@ -191,7 +206,7 @@ public void doGenerate(List<FederateInstance> federates, RtiConfig rtiConfig) {
                 "echo \""
                     + ANSI_INFO
-                    + "RTI has exited. Wait for federates to exit."
+                    + "All components launched. Waiting for federates to exit..."
```
This is confusing. Why are we waiting for federates to exit after all components have launched? Shouldn't we continue?

Jakio815 replied:
I agree, it looked confusing. This part was where the launch script is waiting for the processes (federates) to finish. The processes were already launched before, I fixed it to

`Waiting for federates to complete...` Does this look better?

Or should I just delete this? Fixed in 4d670797a32bb70776783284c210cd08517c6e83.

hokeun commented on core/src/main/java/org/lflang/federated/launcher/FedLauncherGenerator.java:
```diff
@@ -322,13 +358,15 @@ private String getSetupCode() {
         "# Use two distinct traps so we can see which signal causes this.",
         "cleanup() {",
         "    if [ \"$EXITED_SUCCESSFULLY\" = true ] ; then",
+        killAuthCommand,
```
This architecture is confusing. Shouldn't this be added only when SST is used?

Jakio815 replied:
`killAuthCommand` is initialized as an empty string ("") by default. So in other modes (TCP, TLS) it will be a blank line.

I done this because this was the most non-invasive way to handle this conditional Auth cleanup logic inside a static `String.join` block.

Jakio815 replied:
As suggested, I updated the generator to gracefully stop the Auth process by sending a 'quit' command instead of directly killing it.

Here is how the architecture works now:

1. When SST mode is enabled, the launcher creates a unique named pipe (FIFO) locally or on the remote machine (`auth_input_${FEDERATION_ID}.fifo`).
2. The standard input of the Auth process is redirected to read from this FIFO.
3. During the `cleanup()` trap phase (triggered on both successful and unsuccessful exits), the launcher writes 'quit' into the FIFO to signal the Auth process to exit gracefully, and then unlinks/removes the FIFO file.
4. The aggressive process `kill` commands have been completely removed.

hokeun replied:
Ok, understood. I understand you will be using the new quit command for terminating Auth. Can you also rename the variable to `quitAuthCommand` or `terminateAuthCommand` rather than `kill`? `kill` sounds like an error exit.

hokeun replied:
Also I suggest you do something like `isSST ? quitAuthCommand : ""` to make the code more readable, rather than implicitly setting quitAuthCommand to an empty string.

Jakio815 replied:
Fixed on 17404ec80ecb749b9715b0a29ee7e21518e34e3d and c133eb7b651577b1d334a1da8e0fde72f3e60872.

hokeun commented on core/src/main/java/org/lflang/federated/launcher/FedLauncherGenerator.java:
```diff
@@ -322,13 +358,15 @@ private String getSetupCode() {
         "        kill ${pids[@]} || true",
         "        printf \"#### Killing RTI %s.\\n\" ${RTI}",
         "        kill ${RTI} || true",
+        killAuthCommand,
```
Also, the indentation seems off.

Jakio815 replied:
The formatting is done by the `./gradlew :core:spotlessApply` command...

Jakio815 committed:
Fix comment on FedGenerator for TLS and SST.

Jakio815 committed:
Minor fix on comments.

Jakio815 committed:
Fix comments for launch script.

Jakio815 committed:
Formating.

Jakio815 committed:
Build sst-c-api from iotauth submodule to avoid version mismatch in the ci workflow.

Jakio815 committed:
Make SSTGenerator and TLSGenerator instantiable with constructor injection

Jakio815 committed:
Update reactor-c

Jakio815 committed:
Remove unnecessary static keywords from helper methods in SSTGenerator

Jakio815 committed:
Formatting.

Jakio815 committed:
Fix on graph generator.

Jakio815 committed:
Terminate Auth process via FIFO named pipe during cleanup

Jakio815 committed:
Rename killAuthCommand to terminateAuthCommand.

Jakio815 committed:
Add conditions on adding terminateAuthCommand.

Jakio815 committed:
Spotless.

Jakio815 committed:
Merge branch 'master' of github.com:lf-lang/lingua-franca into sst

hokeun commented on core/src/main/java/org/lflang/federated/generator/SSTGenerator.java:
```diff
@@ -0,0 +1,826 @@
+      throw new RuntimeException(e);
+    }
+
+    // Generate SST config for the rti.
```
Nit: rti -> RTI

Jakio815 replied:
Fixed 6357e31dff7379af1616c2ea90d2d55b63552345 with the comment below.

hokeun commented on core/src/main/java/org/lflang/federated/generator/SSTGenerator.java:
```diff
@@ -0,0 +1,826 @@
+
+  public static Path getSSTConfig(FederationFileConfig fileConfig, String name) {
+    return fileConfig.getSSTConfigPath().resolve(name + ".config");
+  }
```
Nit: should be "getSSTConfigPath" because the return type is the path.

Jakio815 replied:
Fixed 6357e31dff7379af1616c2ea90d2d55b63552345

hokeun reviewed (approved):
Looks good to me. I left some additional but minor comments. Please make sure to go over all comments and see if there are any missing/incorrect comments. Please get approval from @edwardalee before merging.

edwardalee commented on test/C/src/federated/DistributedCountTLS.lf:
```diff
@@ -0,0 +1,46 @@
+      lf_print_error_and_exit("Expected to receive 5 items. Received %d.", self->c - 1);
+    }
+  =}
+}
```
```suggestion
import Print from "DistributedCount.lf"
```

edwardalee commented on test/C/src/federated/DistributedCountDecentralizedTLS.lf:
```diff
@@ -0,0 +1,60 @@
+    }
+    printf("SUCCESS: Successfully received 5 items.\n");
+  =}
+}
```
```suggestion
import Print from "DistributedCountDecentralized.lf"
```

edwardalee commented on test/C/src/federated/sst/DistributedCountSSTPermanentDistKey.lf:
```diff
@@ -0,0 +1,52 @@
+      lf_print_error_and_exit("Expected to receive 5 items. Received %d.", self->c - 1);
+    }
+  =}
+}
```
```suggestion
import Print from "../DistributedCount.lf"
```

edwardalee commented on test/C/src/federated/sst/DistributedCountSST.lf:
```diff
@@ -0,0 +1,48 @@
+      lf_print_error_and_exit("Expected to receive 5 items. Received %d.", self->c - 1);
+    }
+  =}
+}
```
```suggestion
import Print from "../DistributedCount.lf"
```

edwardalee commented on test/C/src/federated/sst/DistributedCountDecentralizedSST.lf:
```diff
@@ -0,0 +1,63 @@
+    }
+    printf("SUCCESS: Successfully received 5 items.\n");
+  =}
+}
```
```suggestion
import Print from "../DistributedCountDecentralized.lf"
```

edwardalee commented on core/src/main/java/org/lflang/federated/launcher/FedLauncherGenerator.java:
```diff
@@ -165,6 +173,7 @@ public void doGenerate(List<FederateInstance> federates, RtiConfig rtiConfig) {
                 getDistCode(rtiConfig.getDirectory(), federate.name, federate.user, federate.host))
             .append("\n");
         shCode
+            // TODO: Need to fix calling rtiConfig.getDirectory() here
```
Unclear "TODO". What is this about?

Jakio815 replied:
Removed this comment and the one below here. b48de9b56d01e1b12c7551c3c97bf0ec171f6b44

edwardalee commented on core/src/main/java/org/lflang/federated/launcher/FedLauncherGenerator.java:
```diff
@@ -492,6 +617,27 @@ private String getDistCode(Path remoteBase, String name, String user, String hos
     String buildShellFileName = "build_" + name + ".sh";
     String tarFileName = name + ".tar.gz";
+
+    // ---- Add this block ----
```
What is this "Add this block" comment for?

Jakio815 replied:
Same as above.

edwardalee reviewed (approved):
LGTM.  Some suggestions included.
Can we add some Python tests to make sure this works with Python federates?
I still need to look at the reactor-c PR.
I also requested a Copilot review, which will undoubtedly raise some issues.

Jakio815 committed:
Fix getSSTConfig function name to getSSTConfigPath.

Jakio815 committed:
Minor fix on duplicate CommunicationModeProperty type.

Jakio815 committed:
Fix federate host in SST graph generation.

Jakio815 commented on .github/workflows/c-tests.yml:
```diff
@@ -54,6 +54,17 @@ jobs:
       - name: Install RTI
         uses: ./.github/actions/install-rti
         if: ${{ runner.os == 'macOS' || runner.os == 'Linux' }}
+      - name: Clone SST repository and C library for SST federated tests
+        run: |
+          git clone https://github.com/iotauth/iotauth.git iotauth
+          echo "SST_ROOT=$GITHUB_WORKSPACE/iotauth" >> $GITHUB_ENV
+          cd iotauth
+          git submodule update --init --recursive
+          cd entity/c
```
Will be done in the next PR as soon as the next iotauth release happens.

Jakio815 committed:
Check LOG_TO_FILE instead of $1 for SST auth launcher logging

Jakio815 commented on core/src/main/java/org/lflang/federated/generator/SSTGenerator.java:
```diff
@@ -0,0 +1,826 @@
+    // Create the configuration content
+    StringBuilder configContent = new StringBuilder();
+    configContent
+        .append("entityInfo.name=")
+        .append(entityName)
+        .append("\n")
+        .append("entityInfo.purpose={\"group\":\"RTI\"}\n")
+        .append("entityInfo.number_key=1\n")
```
Intended.

Jakio815 committed:
Update checking for requiring openssl for the cmakegenerator.

Jakio815 commented on core/src/main/java/org/lflang/target/Target.java:
```diff
@@ -550,8 +550,11 @@ public void initialize(TargetConfig config) {
+              SSTProperty.INSTANCE,
               CmakeInitIncludeProperty.INSTANCE,
               CommunicationModeProperty.INSTANCE,
+              SSTProperty.INSTANCE,
```
Fixed 64d48c1aedc98c9d60e55d301b65f1ba9b9c3456.

Jakio815 commented on core/src/main/java/org/lflang/federated/generator/SSTGenerator.java:
```diff
@@ -0,0 +1,826 @@
+    rti.addProperty("host", rtiConfig.getHost());
+    entityList.add(rti);
+
+    // Federate entities
+    for (FederateInstance federate : federateInstances) {
+      String federateName = federate.name;
+      JsonObject entity =
+          createEntity(
+              "Federates", "net1." + federateName, "Net1." + federateName, usePermanentDistKey);
+      entityList.add(entity);
+    }
```
Fixed 23e80ce255d01a59454893cdd8e77c45aadad5a1

Jakio815 commented on core/src/main/java/org/lflang/federated/launcher/FedLauncherGenerator.java:
```diff
@@ -269,82 +284,144 @@ public void doGenerate(List<FederateInstance> federates, RtiConfig rtiConfig) {
+        "\n# Prompt for the password before starting SST Auth",
+        "echo \"Executing Auth.\"",
+        "# Launch the SST Auth.",
+        "if [ \"$1\" = \"-l\" ]; then",
```
Fixed 1e3c2fee7acf39411dbd148e51b35689efdd6685

Jakio815 commented on core/src/main/java/org/lflang/generator/c/CCmakeGenerator.java:
```diff
@@ -413,11 +414,28 @@ CodeBuilder generateCMakeCode(
+        cMakeCode.pr("set(COMM_TYPE " + targetConfig.get(CommunicationModeProperty.INSTANCE) + ")");
+        cMakeCode.newLine();
+      }
+      if (targetConfig.get(CommunicationModeProperty.INSTANCE) == CommunicationMode.SST) {
+        // If security is requested, add the auth option.
+        cMakeCode.pr("# Find OpenSSL and link to it");
+        cMakeCode.pr("find_package(OpenSSL REQUIRED)");
+        cMakeCode.pr("target_link_libraries( ${LF_MAIN_TARGET} PRIVATE OpenSSL::SSL)");
+        cMakeCode.newLine();
+        // If communication mode is SST, find sst package.
+        cMakeCode.pr("# sst-c-api is already linked transitively via reactor-c.");
+      } else if (targetConfig.get(CommunicationModeProperty.INSTANCE) == CommunicationMode.TLS) {
+        // TLS requires OpenSSL only
+        cMakeCode.pr("# Find OpenSSL for TLS support");
+        cMakeCode.pr("find_package(OpenSSL REQUIRED)");
+        cMakeCode.pr(
+            "target_link_libraries(${LF_MAIN_TARGET} PRIVATE OpenSSL::SSL OpenSSL::Crypto)");
+        cMakeCode.newLine();
+      }
```
Fixed here. a20bca00bc7cc01dbdff4f39e73841eac6d67c90

Updated the whole check part of setting openssl_root_dir, when mac, including intel mac and apple mac.

Jakio815 committed:
Import Print reactor as a library from the original DistributedCount.lf to reduce code duplication.

Co-authored-by: Edward A. Lee <eal@eecs.berkeley.edu>

Jakio815 committed:
Remove unnecessary comments and spotless formatting.

Jakio815 committed:
Merge branch 'sst' of github.com:lf-lang/lingua-franca into sst

Jakio815 committed:
Add python examples for SST and TLS, also fix the py-tests to install required libraries such as the c-tests.yml.

Jakio815 commented:
I verified if TLS and SST mode was actually encrypting the communication.

I tested using the federation_id sent in plaintext in TCP mode.

<img width="1667" height="802" alt="image" src="https://github.com/user-attachments/assets/05e7a6eb-998b-428a-8b6e-3b9a22ce7a3f" />
You can check that it shows in plaintext as "test_federation_id_if_plaintext"

However, when I use SST or TLS mode,
<img width="1657" height="792" alt="image" src="https://github.com/user-attachments/assets/54f1c42d-9985-4f44-a7e8-1048f942cd24" />
All bytes are encrypted, and you can also check that the length of the payload is different showing in "Len".

Jakio815 committed:
Formatting...

Jakio815 committed:
Update reactor-c

Jakio815 commented:
> LGTM. Some suggestions included. Can we add some Python tests to make sure this works with Python federates? I still need to look at the reactor-c PR. I also requested a Copilot review, which will undoubtedly raise some issues.

@edwardalee 
I added two more tests in the python test. It's just the same `DistributedCountSST.lf` and `DistributedCountTLS.lf` just python versions.

I added a block for the `python-tests.yml` to install required libraries.

Changes are ab48f55e1879f5b4142fc780304988180b783012.

Jakio815 committed:
Formatting.

Jakio815 committed:
Merge branch 'master' of github.com:lf-lang/lingua-franca into sst

Jakio815 committed:
Update reactor-c
